File: Predicting Type 2 Diabetes Risk from Behavioral and Health Indicator/scripts/python/ml_analysis.py
"""
Machine Learning Analysis
Diabetes Binary Health Indicators - BRFSS 2021
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score, roc_curve
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
import warnings
warnings.filterwarnings('ignore')

# Set style
sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (12, 8)

def load_data():
    """Load and prepare data"""
    df = pd.read_csv('../../data/diabetes_binary_health_indicators_BRFSS2021.csv')
    return df

def prepare_data(df):
    """Prepare data for machine learning"""
    # Separate features and target
    X = df.drop('Diabetes_binary', axis=1)
    y = df['Diabetes_binary']
    
    # Split data
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=y
    )
    
    # Scale features
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)
    
    return X_train, X_test, X_train_scaled, X_test_scaled, y_train, y_test, scaler

def train_models(X_train, X_test, X_train_scaled, X_test_scaled, y_train, y_test):
    """Train multiple machine learning models"""
    print("=" * 60)
    print("TRAINING MACHINE LEARNING MODELS")
    print("=" * 60)
    
    models = {
        'Logistic Regression': LogisticRegression(random_state=42, max_iter=1000),
        'Random Forest': RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1),
        'Gradient Boosting': GradientBoostingClassifier(n_estimators=100, random_state=42),
        'SVM': SVC(probability=True, random_state=42)
    }
    
    results = {}
    
    for name, model in models.items():
        print(f"\n{'='*60}")
        print(f"Training {name}")
        print(f"{'='*60}")
        
        # Use scaled data for Logistic Regression and SVM
        if name in ['Logistic Regression', 'SVM']:
            model.fit(X_train_scaled, y_train)
            y_pred = model.predict(X_test_scaled)
            y_pred_proba = model.predict_proba(X_test_scaled)[:, 1]
        else:
            model.fit(X_train, y_train)
            y_pred = model.predict(X_test)
            y_pred_proba = model.predict_proba(X_test)[:, 1]
        
        # Calculate metrics
        accuracy = accuracy_score(y_test, y_pred)
        precision = precision_score(y_test, y_pred)
        recall = recall_score(y_test, y_pred)
        f1 = f1_score(y_test, y_pred)
        auc = roc_auc_score(y_test, y_pred_proba)
        
        results[name] = {
            'model': model,
            'accuracy': accuracy,
            'precision': precision,
            'recall': recall,
            'f1': f1,
            'auc': auc,
            'y_pred': y_pred,
            'y_pred_proba': y_pred_proba
        }
        
        print(f"Accuracy: {accuracy:.4f}")
        print(f"Precision: {precision:.4f}")
        print(f"Recall: {recall:.4f}")
        print(f"F1 Score: {f1:.4f}")
        print(f"AUC-ROC: {auc:.4f}")
        
        # Classification report
        print(f"\nClassification Report:")
        print(classification_report(y_test, y_pred))
    
    return results

def plot_results(results, y_test, output_dir='../../results/figures/'):
    """Plot model results"""
    # ROC Curves
    plt.figure(figsize=(10, 8))
    for name, result in results.items():
        fpr, tpr, _ = roc_curve(y_test, result['y_pred_proba'])
        plt.plot(fpr, tpr, label=f"{name} (AUC = {result['auc']:.3f})")
    
    plt.plot([0, 1], [0, 1], 'k--', label='Random')
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate')
    plt.ylabel('True Positive Rate')
    plt.title('ROC Curves - Model Comparison', fontsize=14, fontweight='bold')
    plt.legend(loc="lower right")
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig(f'{output_dir}roc_curves.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # Model comparison
    metrics_df = pd.DataFrame({
        name: {
            'Accuracy': result['accuracy'],
            'Precision': result['precision'],
            'Recall': result['recall'],
            'F1 Score': result['f1'],
            'AUC-ROC': result['auc']
        }
        for name, result in results.items()
    }).T
    
    # Visualization
    fig, axes = plt.subplots(1, 2, figsize=(14, 6))
    
    metrics_df[['Accuracy', 'Precision', 'Recall', 'F1 Score']].plot(kind='bar', ax=axes[0])
    axes[0].set_title('Model Performance Metrics', fontsize=12, fontweight='bold')
    axes[0].set_ylabel('Score')
    axes[0].set_xlabel('Model')
    axes[0].legend(loc='upper right')
    axes[0].tick_params(axis='x', rotation=45)
    axes[0].grid(True, alpha=0.3, axis='y')
    
    metrics_df['AUC-ROC'].plot(kind='bar', ax=axes[1], color='steelblue')
    axes[1].set_title('AUC-ROC Scores', fontsize=12, fontweight='bold')
    axes[1].set_ylabel('AUC-ROC')
    axes[1].set_xlabel('Model')
    axes[1].tick_params(axis='x', rotation=45)
    axes[1].grid(True, alpha=0.3, axis='y')
    
    plt.tight_layout()
    plt.savefig(f'{output_dir}model_comparison.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    return metrics_df

def feature_importance_analysis(results, X_train, output_dir='../../results/figures/'):
    """Analyze feature importance for tree-based models"""
    print("=" * 60)
    print("FEATURE IMPORTANCE ANALYSIS")
    print("=" * 60)
    
    # Random Forest feature importance
    if 'Random Forest' in results:
        rf_model = results['Random Forest']['model']
        feature_importance = pd.DataFrame({
            'feature': X_train.columns,
            'importance': rf_model.feature_importances_
        }).sort_values('importance', ascending=False)
        
        print("\nTop 10 Most Important Features (Random Forest):")
        print(feature_importance.head(10))
        
        # Visualization
        plt.figure(figsize=(10, 8))
        top_features = feature_importance.head(15)
        plt.barh(range(len(top_features)), top_features['importance'], color='steelblue')
        plt.yticks(range(len(top_features)), top_features['feature'])
        plt.xlabel('Importance')
        plt.title('Top 15 Feature Importance (Random Forest)', fontsize=14, fontweight='bold')
        plt.gca().invert_yaxis()
        plt.tight_layout()
        plt.savefig(f'{output_dir}feature_importance_rf.png', dpi=300, bbox_inches='tight')
        plt.close()
    
    # Gradient Boosting feature importance
    if 'Gradient Boosting' in results:
        gb_model = results['Gradient Boosting']['model']
        feature_importance_gb = pd.DataFrame({
            'feature': X_train.columns,
            'importance': gb_model.feature_importances_
        }).sort_values('importance', ascending=False)
        
        print("\nTop 10 Most Important Features (Gradient Boosting):")
        print(feature_importance_gb.head(10))
        
        # Visualization
        plt.figure(figsize=(10, 8))
        top_features_gb = feature_importance_gb.head(15)
        plt.barh(range(len(top_features_gb)), top_features_gb['importance'], color='salmon')
        plt.yticks(range(len(top_features_gb)), top_features_gb['feature'])
        plt.xlabel('Importance')
        plt.title('Top 15 Feature Importance (Gradient Boosting)', fontsize=14, fontweight='bold')
        plt.gca().invert_yaxis()
        plt.tight_layout()
        plt.savefig(f'{output_dir}feature_importance_gb.png', dpi=300, bbox_inches='tight')
        plt.close()

def confusion_matrix_plots(results, y_test, output_dir='../../results/figures/'):
    """Plot confusion matrices for all models"""
    fig, axes = plt.subplots(2, 2, figsize=(12, 10))
    axes = axes.ravel()
    
    for idx, (name, result) in enumerate(results.items()):
        cm = confusion_matrix(y_test, result['y_pred'])
        sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', ax=axes[idx])
        axes[idx].set_title(f'Confusion Matrix - {name}', fontsize=12, fontweight='bold')
        axes[idx].set_ylabel('True Label')
        axes[idx].set_xlabel('Predicted Label')
    
    plt.tight_layout()
    plt.savefig(f'{output_dir}confusion_matrices.png', dpi=300, bbox_inches='tight')
    plt.close()

def main():
    """Main function"""
    # Load data
    df = load_data()
    print(f"Dataset shape: {df.shape}")
    
    # Prepare data
    X_train, X_test, X_train_scaled, X_test_scaled, y_train, y_test, scaler = prepare_data(df)
    
    # Train models
    results = train_models(X_train, X_test, X_train_scaled, X_test_scaled, y_train, y_test)
    
    # Plot results
    metrics_df = plot_results(results, y_test)
    
    # Feature importance
    feature_importance_analysis(results, X_train)
    
    # Confusion matrices
    confusion_matrix_plots(results, y_test)
    
    # Save results
    metrics_df.to_csv('../../results/models/model_performance.csv')
    
    print("\n" + "=" * 60)
    print("MACHINE LEARNING ANALYSIS COMPLETE")
    print("=" * 60)
    print("\nBest Model (by AUC-ROC):")
    best_model = metrics_df['AUC-ROC'].idxmax()
    print(f"{best_model}: {metrics_df.loc[best_model, 'AUC-ROC']:.4f}")

if __name__ == "__main__":
    main()

File: Predicting Type 2 Diabetes Risk from Behavioral and Health Indicator/scripts/r/ml_analysis.R
# Machine Learning Analysis
# Diabetes Binary Health Indicators - BRFSS 2021

# Load required libraries
library(caret)
library(randomForest)
library(e1071)
library(pROC)
library(ggplot2)
library(dplyr)

# Set working directory
setwd("../../")

# Load data
df <- read.csv("data/diabetes_binary_health_indicators_BRFSS2021.csv")

cat("Dataset shape:", dim(df), "\n")

# Prepare data
set.seed(42)
trainIndex <- createDataPartition(df$Diabetes_binary, p = 0.8, list = FALSE)
train_data <- df[trainIndex, ]
test_data <- df[-trainIndex, ]

# Separate features and target
X_train <- train_data[, !colnames(train_data) %in% "Diabetes_binary"]
y_train <- train_data$Diabetes_binary
X_test <- test_data[, !colnames(test_data) %in% "Diabetes_binary"]
y_test <- test_data$Diabetes_binary

cat("Training set size:", dim(train_data), "\n")
cat("Test set size:", dim(test_data), "\n")

# Scale features
preProc <- preProcess(X_train, method = c("center", "scale"))
X_train_scaled <- predict(preProc, X_train)
X_test_scaled <- predict(preProc, X_test)

# Train models
cat("\n=== TRAINING MODELS ===\n")

# Logistic Regression
cat("\nTraining Logistic Regression...\n")
log_model <- glm(Diabetes_binary ~ ., data = train_data, family = binomial)
log_pred <- predict(log_model, newdata = test_data, type = "response")
log_pred_class <- ifelse(log_pred > 0.5, 1, 0)

log_accuracy <- mean(log_pred_class == y_test)
log_precision <- sum(log_pred_class == 1 & y_test == 1) / sum(log_pred_class == 1)
log_recall <- sum(log_pred_class == 1 & y_test == 1) / sum(y_test == 1)
log_f1 <- 2 * (log_precision * log_recall) / (log_precision + log_recall)
log_auc <- auc(roc(y_test, log_pred))

cat("Logistic Regression - Accuracy:", log_accuracy, "AUC:", log_auc, "\n")

# Random Forest
cat("\nTraining Random Forest...\n")
rf_model <- randomForest(as.factor(Diabetes_binary) ~ ., data = train_data, ntree = 100, mtry = sqrt(ncol(X_train)))
rf_pred <- predict(rf_model, newdata = test_data, type = "prob")[, 2]
rf_pred_class <- predict(rf_model, newdata = test_data)

rf_accuracy <- mean(rf_pred_class == y_test)
rf_precision <- sum(rf_pred_class == 1 & y_test == 1) / sum(rf_pred_class == 1)
rf_recall <- sum(rf_pred_class == 1 & y_test == 1) / sum(y_test == 1)
rf_f1 <- 2 * (rf_precision * rf_recall) / (rf_precision + rf_recall)
rf_auc <- auc(roc(y_test, rf_pred))

cat("Random Forest - Accuracy:", rf_accuracy, "AUC:", rf_auc, "\n")

# SVM
cat("\nTraining SVM...\n")
svm_model <- svm(as.factor(Diabetes_binary) ~ ., data = train_data, probability = TRUE)
svm_pred <- attr(predict(svm_model, newdata = test_data, probability = TRUE), "probabilities")[, 2]
svm_pred_class <- predict(svm_model, newdata = test_data)

svm_accuracy <- mean(svm_pred_class == y_test)
svm_precision <- sum(svm_pred_class == 1 & y_test == 1) / sum(svm_pred_class == 1)
svm_recall <- sum(svm_pred_class == 1 & y_test == 1) / sum(y_test == 1)
svm_f1 <- 2 * (svm_precision * svm_recall) / (svm_precision + svm_recall)
svm_auc <- auc(roc(y_test, svm_pred))

cat("SVM - Accuracy:", svm_accuracy, "AUC:", svm_auc, "\n")

# Model Comparison
results <- data.frame(
  Model = c("Logistic Regression", "Random Forest", "SVM"),
  Accuracy = c(log_accuracy, rf_accuracy, svm_accuracy),
  Precision = c(log_precision, rf_precision, svm_precision),
  Recall = c(log_recall, rf_recall, svm_recall),
  F1_Score = c(log_f1, rf_f1, svm_f1),
  AUC = c(log_auc, rf_auc, svm_auc)
)

print(results)

# Save results
write.csv(results, "results/models/model_performance_r.csv", row.names = FALSE)

# ROC Curves
png("results/figures/roc_curves_r.png", width = 1000, height = 800, res = 300)
plot(roc(y_test, log_pred), col = "blue", main = "ROC Curves - Model Comparison")
lines(roc(y_test, rf_pred), col = "red")
lines(roc(y_test, svm_pred), col = "green")
legend("bottomright", legend = c("Logistic Regression", "Random Forest", "SVM"),
       col = c("blue", "red", "green"), lty = 1)
dev.off()

# Feature Importance (Random Forest)
if (exists("rf_model")) {
  feature_importance <- importance(rf_model)
  feature_importance_df <- data.frame(
    Feature = rownames(feature_importance),
    Importance = feature_importance[, 1]
  )
  feature_importance_df <- feature_importance_df[order(-feature_importance_df$Importance), ]
  
  cat("\nTop 10 Most Important Features (Random Forest):\n")
  print(head(feature_importance_df, 10))
  
  # Visualization
  png("results/figures/feature_importance_rf_r.png", width = 1000, height = 800, res = 300)
  barplot(head(feature_importance_df$Importance, 15), 
          names.arg = head(feature_importance_df$Feature, 15),
          main = "Top 15 Feature Importance (Random Forest)",
          xlab = "Importance", las = 2, col = "steelblue")
  dev.off()
}

cat("\n=== MACHINE LEARNING ANALYSIS COMPLETE ===\n")
cat("Best Model (by AUC):", results$Model[which.max(results$AUC)], "\n")

